### PR TITLE
Fix jitter in root cert rotator and add integration tests for auto root cert rotation

### PR DESCRIPTION
**Pull request by JimmyCYJ**
Please provide a description for what this PR is for.
Let jitter chooses random seeds to generate backoff time.
Add integration tests to verify auto root cert rotation (https://github.com/istio/istio/pull/17330).

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

https://github.com/istio/istio/issues/17059

### CODE REVIEW

File: tests/integration/security/citadel_root_cert_auto_rotation/main_test.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+			}); err != nil {
+				return err
+			}
+			return nil
+		}).
+		Run()
+
+}
+
+func setupConfig(cfg *istio.Config) {
```
Comment (howardjohn) [inline tests/integration/security/citadel_root_cert_auto_rotation/main_test.go]: can you make this share the suite with the other file_mount one you are adding?

File: tests/integration/security/citadel_root_cert_auto_rotation/main_test.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	g    galley.Instance
+	p    pilot.Instance
+)
+
+func TestMain(m *testing.M) {
+	// Integration test for the Citadel auto root cert upgrade.
+	framework.
+		NewSuite("citadel_root_upgrade_test", m).
+		Label(label.CustomSetup).
+		Label(label.Flaky).
```
Comment (howardjohn) [inline tests/integration/security/citadel_root_cert_auto_rotation/main_test.go]: its flaky already?

File: tests/integration/security/citadel_root_cert_auto_rotation/main_test.go (hunk 1 above)
Comment (JimmyCYJ) [reply tests/integration/security/citadel_root_cert_auto_rotation/main_test.go]: Sure. Done.

File: tests/integration/security/citadel_root_cert_auto_rotation/main_test.go (hunk 2 above)
Comment (JimmyCYJ) [reply tests/integration/security/citadel_root_cert_auto_rotation/main_test.go]: Sorry, it is not flaky. This label is removed.

File: security/cmd/istio_ca/main.go [hunk 3]
```diff
@@ -68,11 +68,11 @@ type cliOptions struct { // nolint: maligned
...
-	selfSignedCA                       bool
-	selfSignedCACertTTL                time.Duration
-	selfSignedRootCertCheckInterval    time.Duration
-	selfSignedRootCertMinCheckInterval time.Duration
-	selfSignedRootCertGracePeriodRatio float64
+	selfSignedCA                       			bool
+	selfSignedCACertTTL                			time.Duration
+	selfSignedRootCertCheckInterval    			time.Duration
+	selfSignedRootCertMinCheckInterval 			time.Duration
+	selfSignedRootCertGracePeriodPercentile int
```
Comment (myidpt) [inline security/cmd/istio_ca/main.go]: lint - more spaces.

**Review by myidpt — COMMENTED** [review]
I think it's better to split the integration test and the real code change into two PRs. WDYT?

File: security/pkg/pki/ca/ca.go [hunk 4]
```diff
@@ -148,12 +148,13 @@ func NewSelfSignedIstioCAOptions(ctx context.Context, readSigningCertOnly bool,
...
 			caCertTTL:           caCertTTL,
 			retryInterval:       cmd.ReadSigningCertCheckInterval,
-			certInspector:       certutil.NewCertUtil(int(rootCertGracePeriodPercentile)),
+			certInspector:       certutil.NewCertUtil(rootCertGracePeriodPercentile),
 			caStorageNamespace:  namespace,
 			dualUse:             dualUse,
 			readSigningCertOnly: readSigningCertOnly,
 			org:                 org,
 			rootCertFile:        rootCertFile,
+			enableJitter:				 enableJitter,
```
Comment (myidpt) [inline security/pkg/pki/ca/ca.go]: lint: spaces

File: security/pkg/pki/ca/selfsignedcarootcertrotator.go [hunk 5]
```diff
@@ -64,11 +67,22 @@ func NewSelfSignedCARootCertRotator(config *SelfSignedCARootCertRotatorConfig,
 		config:              config,
 		ca:                  ca,
 	}
+	if config.enableJitter {
+		// Select a back off time in seconds, which is in the range of [0, rotator.config.CheckInterval).
+		rotator.backOffTime = time.Duration(rand.Int63n(int64(rotator.config.CheckInterval.Seconds())))
```
Comment (myidpt) [inline security/pkg/pki/ca/selfsignedcarootcertrotator.go]: Add a log?

File: security/cmd/istio_ca/main.go [hunk 6]
```diff
@@ -120,6 +132,30 @@ var (
...
+		selfSignedRootCertMinCheckInterval: env.RegisterDurationVar(selfSignedRootCertMinCheckInterval,
+			cmd.SelfSignedRootCertMinCheckInterval,
+			"The minimum interval that self-signed CA checks its root certificate "+
+				"expiration time and rotates root certificate.").Get(),
+		selfSignedRootCertGracePeriodPercentile: env.RegisterIntVar(selfSignedRootCertGracePeriodPercentile,
+			cmd.DefaultRootCertGracePeriodPercentile,
+			"Grace period percentile for self-signed root cert.").Get(),
+		workloadCertMinGracePeriod: env.RegisterDurationVar(workloadCertMinGracePeriod,
+			cmd.DefaultWorkloadMinCertGracePeriod,
+			"The minimum workload certificate rotation grace period.").Get(),
```
Comment (myidpt) [inline security/cmd/istio_ca/main.go]: Why do you add these? Can they be separated into another PR?

File: security/cmd/istio_ca/main.go [hunk 7]
```diff
@@ -192,11 +228,6 @@ func initCLI() {
 	flags.BoolVar(&opts.selfSignedCA, "self-signed-ca", false,
 		"Indicates whether to use auto-generated self-signed CA certificate. "+
 			"When set to true, the '--signing-cert' and '--signing-key' options are ignored.")
-	flags.DurationVar(&opts.selfSignedCACertTTL, "self-signed-ca-cert-ttl", cmd.DefaultSelfSignedCACertTTL,
-		"The TTL of self-signed CA root certificate.")
-	flags.DurationVar(&opts.selfSignedRootCertCheckInterval, "self-signed-root-check-interval", cmd.DefaultSelfSignedRootCertCheckInterval,
-		"The interval that self-signed CA checks its root certificate expiration time and rotates root certificate. "+
-			"Should not be shorter than one minute.")
```
Comment (myidpt) [inline security/cmd/istio_ca/main.go]: If you change these, you will need to change the helm templates. Please only change what is needed in this PR :)

Comment (myidpt) [reply security/cmd/istio_ca/main.go]: Never mind. It seems these flags are not exposed in helm.

File: security/cmd/istio_ca/main.go [hunk 8]
```diff
@@ -47,6 +47,15 @@ import (
 	"istio.io/pkg/version"
 )
 
+const (
+	selfSignedCaCertTTL                     = "SELF_SIGNED_CA_CERT_TTL"
+	selfSignedRootCertCheckInterval         = "SELF_SIGNED_ROOT_CERT_CHECK_INTERVAL"
+	selfSignedRootCertMinCheckInterval      = "SELF_SIGNED_ROOT_CERT_MIN_CHECK_INTERVAL"
```
Comment (myidpt) [inline security/cmd/istio_ca/main.go]: Can we make this a constant?

**Comment by JimmyCYJ** [comment]
@myidpt The jitter change is in https://github.com/istio/istio/pull/17597 now. Thanks.

File: security/cmd/istio_ca/main.go [hunk 9]
```diff
@@ -64,10 +69,12 @@ type cliOptions struct { // nolint: maligned
...
 	rootCertFile    string
 
-	selfSignedCA                    bool
-	selfSignedCACertTTL             time.Duration
-	selfSignedRootCertCheckInterval time.Duration
-	enableJitterForRootCertRotator  bool
+	selfSignedCA                            bool
+	selfSignedCACertTTL                     time.Duration
+	selfSignedRootCertCheckInterval         time.Duration
+	selfSignedRootCertMinCheckInterval      time.Duration
```
Comment (myidpt) [inline security/cmd/istio_ca/main.go]: Do we need this? Can this be a constant?

File: security/pkg/pki/ca/selfsignedcarootcertrotator.go [hunk 10]
```diff
@@ -69,7 +69,9 @@ func NewSelfSignedCARootCertRotator(config *SelfSignedCARootCertRotatorConfig,
 	}
 	if config.enableJitter {
 		// Select a back off time in seconds, which is in the range of [0, rotator.config.CheckInterval).
-		backOffSeconds := int(time.Duration(rand.Int63n(int64(rotator.config.CheckInterval))).Seconds())
+		randSource := rand.NewSource(time.Now().UnixNano())
+		randBackOff := rand.New(randSource)
+		backOffSeconds := int(time.Duration(randBackOff.Int63n(int64(rotator.config.CheckInterval))).Seconds())
```
Comment (myidpt) [inline security/pkg/pki/ca/selfsignedcarootcertrotator.go]: Good call. The random source should be set.

**Review by myidpt — COMMENTED** [review]
Looks good to me. Please fix the comments and tests.

File: security/cmd/istio_ca/main.go (hunk 9 above)
Comment (JimmyCYJ) [reply security/cmd/istio_ca/main.go]: We need to change these values in our integration tests and perf tests. The default CACertTTL is 10 years, and check interval is one hour.

**Review by myidpt — APPROVED** [review]
Thanks for the work!

File: install/kubernetes/helm/istio/charts/security/templates/deployment.yaml [hunk 11]
```diff
@@ -88,6 +88,12 @@ spec:
             initialDelaySeconds: 60
             periodSeconds: 60
           {{- end }}
+            {{- if .Values.env }}
```
Comment (howardjohn) [inline install/kubernetes/helm/istio/charts/security/templates/deployment.yaml]: can you send a PR to istio/installer

File: install/kubernetes/helm/istio/charts/security/values.yaml [hunk 12]
```diff
@@ -18,6 +18,8 @@ citadelHealthCheck: false
 # 90*24hour = 2160h
 workloadCertTtl: 2160h
 
+env: {}
```
Comment (howardjohn) [inline install/kubernetes/helm/istio/charts/security/values.yaml]: Can you add a comment? this shows up in user docs on istio.io

File: tests/integration/security/file_mount_mtls_root_cert_auto_rotation/multiple_citadel_root_rotation_test.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+					ExpectSuccess: true,
+				},
+			}
+			for _, checker := range checkers {
+				retry.UntilSuccessOrFail(t, checker.Check, retry.Delay(time.Second), retry.Timeout(10*time.Second))
+			}
+
+			// Wait for at least one root upgrade to let workloads use new CA cert
+			// to set up mTLS connections.
+			time.Sleep(1 * time.Minute)
```
Comment (howardjohn) [inline tests/integration/security/file_mount_mtls_root_cert_auto_rotation/multiple_citadel_root_rotation_test.go]: wow this is super long !! is there any way we can poll for this instead of waiting 1 min? See https://github.com/istio/istio/issues/17512, security is already by far the longest test suite

File: tests/integration/security/file_mount_mtls_root_cert_auto_rotation/multiple_citadel_root_rotation_test.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+			for i := 0; i < 5; i++ {
+				for _, checker := range checkers {
+					retry.UntilSuccessOrFail(t, checker.Check, retry.Delay(time.Second), retry.Timeout(10*time.Second))
+				}
+				time.Sleep(time.Duration(requestIntervalSeconds) * time.Second)
```
Comment (howardjohn) [inline tests/integration/security/file_mount_mtls_root_cert_auto_rotation/multiple_citadel_root_rotation_test.go]: Why do we have to sleep here? This is waiting another minute

File: tests/integration/security/file_mount_mtls_root_cert_auto_rotation/multiple_citadel_root_rotation_test.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+				},
+			}
+			for _, checker := range checkers {
+				retry.UntilSuccessOrFail(t, checker.Check, retry.Delay(time.Second), retry.Timeout(10*time.Second))
+			}
+
+			// Wait for at least one root upgrade to let workloads use new CA cert
+			// to set up mTLS connections.
+			time.Sleep(1 * time.Minute)
+			for i := 0; i < 5; i++ {
```
Comment (howardjohn) [inline tests/integration/security/file_mount_mtls_root_cert_auto_rotation/multiple_citadel_root_rotation_test.go]: why loop here?

File: tests/integration/security/file_mount_mtls_root_cert_auto_rotation/file_mount_mtls_test.go [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+					ExpectSuccess: true,
+				},
+			}
+			for _, checker := range checkers {
+				retry.UntilSuccessOrFail(t, checker.Check, retry.Delay(time.Second), retry.Timeout(10*time.Second))
+			}
+
+			// Wait for at least one root upgrade to let workloads use new CA cert
+			// to set up mTLS connections.
+			time.Sleep(1 * time.Minute)
```
Comment (howardjohn) [inline tests/integration/security/file_mount_mtls_root_cert_auto_rotation/file_mount_mtls_test.go]: same comment as before

File: tests/integration/security/file_mount_mtls_root_cert_auto_rotation/file_mount_mtls_test.go [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+			for _, checker := range checkers {
```
Comment (howardjohn) [inline tests/integration/security/file_mount_mtls_root_cert_auto_rotation/file_mount_mtls_test.go]: why do we loop 5 times below and not here?

File: tests/integration/security/file_mount_mtls_root_cert_auto_rotation/citadel_root_cert_auto_rotation_test.go [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+			// Get initial root cert.
+			kubeAccessor := ctx.Environment().(*kube.Environment).Accessor
+			systemNS := namespace.ClaimOrFail(t, ctx, istioCfg.SystemNamespace)
+			caScrt, err := kubeAccessor.GetSecret(systemNS.Name()).Get(CASecret, metav1.GetOptions{})
+			if err != nil {
+				t.Fatalf("unable to load root secret: %s", err.Error())
+			}
+
+			// Wait for the next round of root cert upgrade
+			time.Sleep(1 * time.Minute)
```
Comment (howardjohn) [inline tests/integration/security/file_mount_mtls_root_cert_auto_rotation/citadel_root_cert_auto_rotation_test.go]: same comment

File: tests/integration/security/file_mount_mtls_root_cert_auto_rotation/file_mount_mtls_test.go (hunk 16 above)
Comment (howardjohn) [reply tests/integration/security/file_mount_mtls_root_cert_auto_rotation/file_mount_mtls_test.go]: actually what are you even testing here... we spin up `echo`, then wait 1 minute... but we didn't do anything. How are we getting the "new ca cert"?

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: install/kubernetes/helm/istio/charts/security/templates/deployment.yaml (hunk 11 above)
Comment (JimmyCYJ) [reply install/kubernetes/helm/istio/charts/security/templates/deployment.yaml]: Sure, I send out https://github.com/istio/installer/pull/422. Please take a look. Thanks!

**Comment by myidpt** [comment]
@howardjohn since the PR istio/installer#422 is already merged, do you have other comments before approving this one?

**Review by howardjohn — APPROVED** [review]
